File: ui/src/main/java/com/doordeck/sdk/ui/unlock/UnlockActivity.java
package com.doordeck.sdk.ui.unlock;

import android.Manifest;
import android.annotation.SuppressLint;
import android.content.DialogInterface;
import android.content.IntentSender;
import android.content.pm.PackageManager;
import android.graphics.Color;
import android.graphics.drawable.Animatable;
import android.graphics.drawable.ColorDrawable;
import android.graphics.drawable.Drawable;
import android.graphics.drawable.GradientDrawable;
import android.graphics.drawable.ShapeDrawable;
import android.os.Build;

import androidx.annotation.NonNull;

import butterknife.ButterKnife;
import com.google.android.material.floatingactionbutton.FloatingActionButton;
import androidx.core.app.ActivityCompat;
import androidx.core.content.ContextCompat;
import androidx.interpolator.view.animation.FastOutSlowInInterpolator;
import androidx.appcompat.app.AlertDialog;
import android.os.Bundle;
import androidx.appcompat.app.AppCompatDelegate;

import android.util.TypedValue;
import android.view.View;
import android.view.animation.OvershootInterpolator;
import android.widget.Button;
import android.widget.ImageView;
import android.widget.ProgressBar;
import android.widget.TextView;

//import com.doordeck.sdk.Doordeck;
import com.doordeck.sdk.R2;
import com.doordeck.sdk.common.contstants.Constants;
import com.doordeck.sdk.common.utils.PreferencesManager;
import com.doordeck.sdk.ui.BaseActivity;
import com.google.android.gms.common.api.ResolvableApiException;
import com.google.android.gms.location.LocationRequest;
import com.google.android.gms.location.LocationServices;
import com.google.android.gms.location.LocationSettingsRequest;
import com.google.android.gms.location.LocationSettingsResponse;
import com.google.android.gms.location.SettingsClient;
import com.google.android.gms.tasks.OnFailureListener;
import com.google.android.gms.tasks.OnSuccessListener;
import com.google.android.gms.tasks.Task;

import butterknife.BindView;
import butterknife.OnClick;

import static com.doordeck.sdk.common.contstants.Constants.prefs.REFRESHTOKEN;
import static com.doordeck.sdk.common.contstants.Constants.prefs.TOKEN;

public class UnlockActivity extends BaseActivity implements UnlockView {
//public class UnlockActivity extends BaseActivity implements UnlockView, Doordeck.AuthenticateCallback {

    static {
        AppCompatDelegate.setCompatVectorFromResourcesEnabled(true);
    }

    @BindView(R2.id.key_title)
    TextView keyTV;

    @BindView(R2.id.unlock_status)
    TextView statusTV;

    @BindView(R2.id.logo_spinner)
    ProgressBar spinnerPB;

    @BindView(R2.id.lock_bkg)
    ImageView circleBkgIV;

    @BindView(R2.id.arrow_image)
    ImageView arrowIV;

    @BindView(R2.id.dismissBtn)
    Button dismissBtn;

    @BindView(R2.id.fav_button)
    FloatingActionButton favBtn;

    @OnClick(R2.id.dismissBtn)
    public void onDismissClicked() {
        finish();
    }

    @OnClick(R2.id.fav_button)
    public void onFavButtonClicked() {
        if (theme == R2.style.ddlightTheme) {
            PreferencesManager.getInstance().setInt(Constants.prefs.LIGHT_THEME, R2.style.ddDarkTheme);
        } else {
            PreferencesManager.getInstance().setInt(Constants.prefs.LIGHT_THEME, R2.style.ddlightTheme);
        }
    }

    private UnlockPresenter unlockPresenter;
    private boolean locationPermissionShown = false;
    private boolean googlePermissionShown = false;
    private int theme;


    @Override
    public void onCreate(Bundle savedInstanceState) {
//
//        Doordeck.initialize(this);
        super.onCreate(savedInstanceState);
        setContentView(R2.layout.activity_unlock);
        ButterKnife.bind(this);

        View decorView = getWindow().getDecorView();
        int uiOptions = View.SYSTEM_UI_FLAG_HIDE_NAVIGATION
                | View.SYSTEM_UI_FLAG_FULLSCREEN | View.SYSTEM_UI_FLAG_IMMERSIVE_STICKY | View.SYSTEM_UI_FLAG_LAYOUT_STABLE | View.SYSTEM_UI_FLAG_LAYOUT_HIDE_NAVIGATION;
        decorView.setSystemUiVisibility(uiOptions);
        if(unlockPresenter == null) {
            unlockPresenter = new UnlockPresenter(this);
        }
        resetAnimation();



    }

//    @Override
    public void authenticationSuccess() {
        if(unlockPresenter == null) {
            unlockPresenter = new UnlockPresenter(this);
        }
        unlockPresenter.handleIntent(getIntent());
    }

//    @Override
    public void authenticationFail(String error) {
        showAuthError(error);
    }

    private void showAuthError(String error) {
        statusTV.setText(error);
        showAccessDeniedAnimation();
    }




    @Override
    public void showNoAccessGeoFence() {
        showAccessDeniedAnimation();
        statusTV.setText(R2.string.ACCESS_DENIED_GEOFENCE);
    }

    @Override
    public void updateLockName(String name) {
        keyTV.setText(name);
    }

    @Override
    public void setUnlocking() {
        statusTV.setText(getText(R2.string.UNLOCKING));
    }

    @Override
    public void showGeoLoading() {
        statusTV.setText(R2.string.CHECKING_GEOFENCE);
    }



    @Override
    public void finishActivity() {
        finish();
    }


    private void resetAnimation() {

        spinnerPB.setVisibility(View.VISIBLE);
        arrowIV.setScaleX(0f);
        arrowIV.setScaleY(0f);
        circleBkgIV.setScaleX(1f);
        circleBkgIV.setScaleY(1f);
        Drawable circle = circleBkgIV.getDrawable();
        TypedValue typedValue = new TypedValue();
        this.getTheme().resolveAttribute(R2.attr.ddColorAccent, typedValue, true);
        int colorAccent = typedValue.data;
        if (circle instanceof ShapeDrawable) {
            ((ShapeDrawable)circle).getPaint().setColor(colorAccent);
        } else if (circle instanceof GradientDrawable) {
            ((GradientDrawable)circle).setColor(colorAccent);
        } else if (circle instanceof ColorDrawable) {
            ((ColorDrawable)circle).setColor(colorAccent);
        }
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            arrowIV.setImageDrawable(ContextCompat.getDrawable(this,R2.drawable.ic_unlock_success));
        }
        else{
            arrowIV.setImageDrawable(ContextCompat.getDrawable(this,R2.drawable.ic_unlock_success));
        }
        statusTV.setText(R2.string.UNLOCKING);
    }

    @Override
    public void unlockSuccess() {
        showUnlockAnimation();
        unlockPresenter.setFinishTimer();
    }


    public void showUnlockAnimation(){

        spinnerPB.setAlpha(0);
        Drawable circle = circleBkgIV.getDrawable();
        if (circle instanceof ShapeDrawable) {
            ((ShapeDrawable)circle).getPaint().setColor(ContextCompat.getColor(this,R2.color.success));
        } else if (circle instanceof GradientDrawable) {
            ((GradientDrawable)circle).setColor(ContextCompat.getColor(this,R2.color.success));
        } else if (circle instanceof ColorDrawable) {
            ((ColorDrawable)circle).setColor(ContextCompat.getColor(this,R2.color.success));
        }
        circleBkgIV.animate().scaleX(25f).scaleY(25f).setInterpolator(new FastOutSlowInInterpolator()).setDuration(500);
        arrowIV.animate().scaleX(1f).scaleY(1f).alpha(1.0f).setInterpolator(new OvershootInterpolator()).setDuration(300).setStartDelay(200);
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            arrowIV.setImageDrawable(ContextCompat.getDrawable(this,R2.drawable.ic_unlock_success));
        }
        else{
            arrowIV.setImageDrawable(ContextCompat.getDrawable(this,R2.drawable.ic_unlock_success));
        }
        Drawable animation = arrowIV.getDrawable();
        if (animation instanceof Animatable) {
            ((Animatable) animation).start();
        }
        statusTV.setText(R2.string.UNLOCKED);
        statusTV.setTextColor(ContextCompat.getColor(this, R2.color.ddColorTextLight));
        favBtn.setAlpha(0.0f);
        favBtn.setVisibility(View.VISIBLE);
        favBtn.animate().alpha(1.0f).setDuration(200);
    }

    @Override
    public void showAccessDenied() {
        showAccessDeniedAnimation();
        statusTV.setText("");
    }

    public void showAccessDeniedAnimation(){

        long duration = 600;
        spinnerPB.setAlpha(0);

        Drawable circle = circleBkgIV.getDrawable();
        if (circle instanceof ShapeDrawable) {
            ((ShapeDrawable)circle).getPaint().setColor(ContextCompat.getColor(this,R2.color.error));
        } else if (circle instanceof GradientDrawable) {
            ((GradientDrawable)circle).setColor(ContextCompat.getColor(this,R2.color.error));
        } else if (circle instanceof ColorDrawable) {
            ((ColorDrawable)circle).setColor(ContextCompat.getColor(this,R2.color.error));
        }
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            arrowIV.setImageDrawable(ContextCompat.getDrawable(this,R2.drawable.ic_access_denied));
        }
        else{
            arrowIV.setImageDrawable(ContextCompat.getDrawable(this,R2.drawable.ic_access_denied));
        }
        circleBkgIV.animate().scaleX(25f).scaleY(25f).setInterpolator(new FastOutSlowInInterpolator()).setDuration(500);
        arrowIV.animate().scaleX(1.2f).scaleY(1.2f).alpha(1.0f).setInterpolator(new OvershootInterpolator()).setDuration(duration);
        Drawable animation = arrowIV.getDrawable();
        if (animation instanceof Animatable) {
            ((Animatable) animation).start();
        }
        keyTV.setText(R2.string.ACCESS_DENIED);
        statusTV.setTextColor(Color.WHITE);

    }

    @Override
    public void checkGoogleApiPermissions() {
        if (checkLocationPermission()){

            LocationRequest mLocationRequest = new LocationRequest();
            mLocationRequest.setPriority(LocationRequest.PRIORITY_HIGH_ACCURACY);
            // check if location is enabled
            LocationSettingsRequest.Builder builder = new LocationSettingsRequest.Builder().addLocationRequest(mLocationRequest);
            SettingsClient client = LocationServices.getSettingsClient(this);
            Task<LocationSettingsResponse> task = client.checkLocationSettings(builder.build());

            // location is enabled
            task.addOnSuccessListener(this, new OnSuccessListener<LocationSettingsResponse>() {
                @SuppressLint("MissingPermission")
                @Override
                public void onSuccess(LocationSettingsResponse locationSettingsResponse) {
                    unlockPresenter.checkGeofence();
                }
            });

            // location is disabled
            task.addOnFailureListener(this, new OnFailureListener() {
                @Override
                public void onFailure(@NonNull Exception e) {
                    if (e instanceof ResolvableApiException) {
                        // LocationObject settings are not satisfied, but this can be fixed
                        // by showing the user a dialog.
                        try {
                            // Show the dialog by calling startResolutionForResult(),
                            // and check the result in onActivityResult().
                            if(!googlePermissionShown) {
                                googlePermissionShown = true;
                                ResolvableApiException resolvable = (ResolvableApiException) e;
                                resolvable.startResolutionForResult(UnlockActivity.this,
                                        Constants.permission.LOCATION);
                            }
                        } catch (IntentSender.SendIntentException sendEx) {
                            // Ignore the error.
                            sendEx.printStackTrace();
                        }
                    } else {
                        e.printStackTrace();
                    }
                }
            });
        }
    }

    public boolean checkLocationPermission() {
        if (Build.VERSION.SDK_INT >= 23 && ContextCompat.checkSelfPermission(this,
                Manifest.permission.ACCESS_FINE_LOCATION)
                != PackageManager.PERMISSION_GRANTED) {

            // Should we show an explanation?
            if (ActivityCompat.shouldShowRequestPermissionRationale(this,
                    Manifest.permission.ACCESS_FINE_LOCATION)) {

                // Show an explanation to the user *asynchronously* -- don't block
                // this thread waiting for the user's response! After the user
                // sees the explanation, try again to request the permission.
                if(!locationPermissionShown) {
                    locationPermissionShown = true;
                    new AlertDialog.Builder(this)
                            .setTitle(R2.string.LOCATION_PERMISSION_TITLE)
                            .setMessage(R2.string.LOCATION_PERMISSION_TEXT)
                            .setPositiveButton(R2.string.OK, new DialogInterface.OnClickListener() {
                                @Override
                                public void onClick(DialogInterface dialogInterface, int i) {
                                    //Prompt the user once explanation has been shown
                                    ActivityCompat.requestPermissions(UnlockActivity.this,
                                            new String[]{Manifest.permission.ACCESS_FINE_LOCATION},
                                            Constants.permission.LOCATION);
                                }
                            })
                            .create()
                            .show();
                }

            } else {
                // No explanation needed, we can request the permission.
                ActivityCompat.requestPermissions(this,
                        new String[]{Manifest.permission.ACCESS_FINE_LOCATION},
                        Constants.permission.LOCATION);
            }
            return false;
        } else {
            return true;
        }
    }



}

File: ui/src/main/java/com/doordeck/sdk/ui/unlock/UnlockActivity.kt
package com.doordeck.sdk.ui.unlock

import android.Manifest
import android.content.Context
import android.content.Intent
import android.content.IntentSender
import android.content.pm.PackageManager
import android.graphics.PorterDuff
import android.graphics.drawable.Animatable
import android.os.Bundle
import android.view.View
import android.view.animation.AccelerateInterpolator
import android.view.animation.OvershootInterpolator
import android.widget.Toast
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatDelegate
import androidx.core.app.ActivityCompat
import androidx.core.content.ContextCompat
import androidx.vectordrawable.graphics.drawable.AnimatedVectorDrawableCompat
import com.doordeck.multiplatform.sdk.model.responses.LocationRequirementResponse
import com.doordeck.multiplatform.sdk.model.responses.LockResponse
import com.doordeck.multiplatform.sdk.model.responses.SiteLocksResponse
import com.doordeck.sdk.common.manager.AuthStatus
import com.doordeck.sdk.common.manager.Doordeck
import com.doordeck.sdk.common.manager.authStatus
import com.doordeck.sdk.common.utils.json
import com.doordeck.sdk.ui.BaseActivity
import com.doordeck.sdk.ui.nfc.NFCActivity
import com.doordeck.sdk.ui.qrcode.QRcodeActivity
import com.doordeck.sdk.ui.showlistofdevicestounlock.ShowListOfDevicesToUnlockActivity
import com.doordeck.sdk.ui.verify.VerifyDeviceActivity
import com.github.doordeck.ui.R
import com.github.doordeck.ui.databinding.ActivityUnlockBinding
import com.google.android.gms.common.api.ResolvableApiException
import com.google.android.gms.location.LocationRequest
import com.google.android.gms.location.LocationServices
import com.google.android.gms.location.LocationSettingsRequest
import com.google.android.gms.location.Priority
import kotlinx.serialization.json.Json

// screen responsible to display the status of the unlock process
internal class UnlockActivity : BaseActivity(), UnlockView {

    private var unlockPresenter: UnlockPresenter? = null
    private var locationPermissionShown = false
    private var googlePermissionShown = false
    private var canceledVerify = false

    // Using this to avoid entering on onResume after a beaming.
    //
    // There can be a situation when the user beams, goes to see result and beams
    // again while/after computing the result without clicking "Dismiss"
    private var unlockFinished = false

    private lateinit var binding: ActivityUnlockBinding

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        binding = ActivityUnlockBinding.inflate(layoutInflater)
        setContentView(binding.root)

        unlockPresenter = UnlockPresenter()
        binding.tvDismiss.setOnClickListener { back() }
    }

    private fun back() {
        finish()
        when (comingFrom) {
            COMING_FROM_DIRECT_UNLOCK -> {
                // NO-OP, just finish ☝️
            }

            COMING_FROM_QR_SCAN -> QRcodeActivity.start(this)
            COMING_FROM_NFC -> NFCActivity.start(this)
        }

        unlockFinished = false
    }

    override fun showNoAccessGeoFence() {
        showAccessDeniedAnimation()
        binding.unlockStatus.setText(R.string.ACCESS_DENIED_GEOFENCE)

        unlockFinished = true
    }

    override fun updateLockName(name: String) {
        binding.keyTitle.text = name
    }

    override fun setUnlocking() {
        binding.unlockStatus.text = getText(R.string.UNLOCKING)
    }

    override fun showGeoLoading() {
        binding.unlockStatus.setText(R.string.CHECKING_GEOFENCE)
    }


    override fun finishActivity() {
        back()
    }


    private fun resetAnimation() {
        binding.circleBack.scaleX = 0f
        binding.circleBack.scaleY = 0f

        binding.keyTitle.alpha = 0f

        val animated = AnimatedVectorDrawableCompat.create(this, R.drawable.ic_unlock_success)
        val animation = binding.lockImage.drawable
        if (animation is Animatable) {
            (animation as Animatable).stop()
        }
        binding.lockImage.setImageDrawable(null)
        binding.lockImage.setImageDrawable(animated)
        binding.logoSpinner.visibility = View.VISIBLE
        binding.unlockStatus.setText(R.string.UNLOCKING)
    }

    override fun unlockSuccess() {
        showUnlockAnimation()
        unlockPresenter?.setFinishTimer()
    }


    private fun showUnlockAnimation() {

        binding.tvDismiss.setBackgroundColor(ContextCompat.getColor(this, R.color.black_transp))
        binding.circleBack.setColorFilter(ContextCompat.getColor(this, R.color.success), PorterDuff.Mode.SRC_IN)
        binding.circleBack.animate().alpha(1f).scaleX(13f).scaleY(13f).setInterpolator(AccelerateInterpolator()).setDuration(500)
        binding.logoSpinner.alpha = 0f
        val animated = AnimatedVectorDrawableCompat.create(this, R.drawable.ic_unlock_success)
        binding.lockImage.setImageDrawable(animated)
        val animation = binding.lockImage.drawable
        if (animation is Animatable) {
            (animation as Animatable).start()
        }
        binding.unlockStatus.setText(R.string.UNLOCKED)
        binding.keyTitle.animate().alpha(1f).translationY(150f).setInterpolator(OvershootInterpolator()).setDuration(500).startDelay = 500
    }

    override fun showAccessDenied() {
        showAccessDeniedAnimation()
        binding.unlockStatus.text = ""

        unlockFinished = true
    }

    override fun notValidTileId() {
        Toast.makeText(this, getString(R.string.tile_id_not_valid), Toast.LENGTH_LONG).show()
        back()
    }

    override fun noUserLoggedIn() {
        showAccessDeniedAnimation()
        binding.unlockStatus.text = getString(R.string.no_user_logged_in)

        unlockFinished = true
    }

    override fun goToDevices(devices: List<LockResponse>) {
        ShowListOfDevicesToUnlockActivity.start(this, devices)
    }

    override fun getDefaultLockColours(): Array<String> {
        return resources.getStringArray(R.array.lock_colours)
    }

    override fun displayVerificationView() {
        if (!canceledVerify) {
            canceledVerify = true
            VerifyDeviceActivity.start(this)
        } else {
            back()
        }
    }

    private fun showAccessDeniedAnimation() {
        binding.tvDismiss.setBackgroundColor(ContextCompat.getColor(this, R.color.black_transp))
        binding.circleBack.setColorFilter(ContextCompat.getColor(this, R.color.error), PorterDuff.Mode.SRC_IN)
        binding.circleBack.animate().alpha(1f).scaleX(13f).scaleY(13f).setInterpolator(AccelerateInterpolator()).setDuration(500)
        binding.logoSpinner.alpha = 0f
        val animated = AnimatedVectorDrawableCompat.create(this, R.drawable.ic_unlock_fail)
        binding.lockImage.setImageDrawable(animated)
        val animation = binding.lockImage.drawable
        if (animation is Animatable) {
            (animation as Animatable).start()
        }
        binding.keyTitle.setText(R.string.ACCESS_DENIED)
        binding.keyTitle.animate().alpha(1f).translationY(150f).setInterpolator(OvershootInterpolator()).setDuration(500).startDelay = 500

        unlockFinished = true
    }

    override fun checkGoogleApiPermissions(device: LockResponse, location: LocationRequirementResponse) {
        if (checkLocationPermission()) {

            val locationRequest = LocationRequest
                .Builder(Priority.PRIORITY_HIGH_ACCURACY, 100)
                .build()
            // check if location is enabled
            val builder = LocationSettingsRequest.Builder().addLocationRequest(locationRequest)
            val client = LocationServices.getSettingsClient(this)
            val task = client.checkLocationSettings(builder.build())

            // location is enabled
            task.addOnSuccessListener(this) { unlockPresenter?.checkGeofence(device, location) }

            // location is disabled
            task.addOnFailureListener(this) { e ->
                if (e is ResolvableApiException) {
                    // LocationObject settings are not satisfied, but this can be fixed
                    // by showing the user a dialog.
                    try {
                        // Show the dialog by calling startResolutionForResult(),
                        // and check the result in onActivityResult().
                        if (!googlePermissionShown) {
                            googlePermissionShown = true
                            e.startResolutionForResult(
                                this@UnlockActivity,
                                LOCATION
                            )
                        }
                    } catch (sendEx: IntentSender.SendIntentException) {
                        // Ignore the error.
                        sendEx.printStackTrace()
                    }

                } else {
                    e.printStackTrace()
                }
            }
        }
    }

    private fun checkLocationPermission(): Boolean {
        if (ContextCompat.checkSelfPermission(
                this,
                Manifest.permission.ACCESS_FINE_LOCATION
            ) != PackageManager.PERMISSION_GRANTED
        ) {

            // Should we show an explanation?
            if (ActivityCompat.shouldShowRequestPermissionRationale(
                    this,
                    Manifest.permission.ACCESS_FINE_LOCATION
                )
            ) {

                // Show an explanation to the user *asynchronously* -- don't block
                // this thread waiting for the user's response! After the user
                // sees the explanation, try again to request the permission.
                if (!locationPermissionShown) {
                    locationPermissionShown = true
                    AlertDialog.Builder(this)
                        .setTitle(R.string.LOCATION_PERMISSION_TITLE)
                        .setMessage(R.string.LOCATION_PERMISSION_TEXT)
                        .setPositiveButton(R.string.OK) { _, _ ->
                            //Prompt the user once explanation has been shown
                            ActivityCompat.requestPermissions(
                                this@UnlockActivity,
                                arrayOf(Manifest.permission.ACCESS_FINE_LOCATION),
                                LOCATION
                            )
                        }
                        .create()
                        .show()
                }

            } else {
                // No explanation needed, we can request the permission.
                ActivityCompat.requestPermissions(
                    this,
                    arrayOf(Manifest.permission.ACCESS_FINE_LOCATION),
                    LOCATION
                )
            }
            return false
        } else {
            return true
        }
    }

    @Suppress("OVERRIDE_DEPRECATION")
    override fun onBackPressed() {
        back()
    }

    public override fun onStart() {
        super.onStart()
        unlockPresenter?.onStart(this)
    }

    public override fun onResume() {
        super.onResume()
        if (unlockFinished) {
            return
        }

        resetAnimation()

        if (Doordeck.getHeadlessInstance(this).authStatus == AuthStatus.UNAUTHORIZED) {
            noUserLoggedIn()
            return
        }
        val tileId = this.tileId
        val deviceJson = this.deviceJson

        if (tileId != null) {
            unlockPresenter?.resolveTile(tileId)
        } else if (deviceJson != null) {
            val deviceToUnlock = json.decodeFromString<SiteLocksResponse>(deviceJson)
            unlockPresenter?.resolveTile(deviceToUnlock.id)
        }
    }

    public override fun onPause() {
        super.onPause()
        unlockPresenter = null
    }

    public override fun onStop() {
        super.onStop()
        finish()
    }

    override fun onDestroy() {
        super.onDestroy()
        unlockPresenter = null
    }

    companion object {

        private const val LOCATION = 99

        init {
            AppCompatDelegate.setCompatVectorFromResourcesEnabled(true)
        }

        private const val TILE_ID = "TILE_ID"
        private const val DEVICE = "DEVICE"


        /**
         * Using this from the QR/NFC viewer, so we finish the activity which is the previous one,
         * if it's an activity
         */
        fun start(context: Context, tileId: String, comingFrom: String) {
            val starter = Intent(context, UnlockActivity::class.java)
            starter.putExtra(TILE_ID, tileId)
            starter.putExtra(COMING_FROM_KEY, comingFrom)
            if (comingFrom != COMING_FROM_DIRECT_UNLOCK) {
                starter.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
            }
            context.startActivity(starter)
        }

        fun start(context: Context, device: LockResponse, comingFrom: String) {
            val starter = Intent(context, UnlockActivity::class.java)
            starter.putExtra(DEVICE, Json.encodeToString(device))
            starter.putExtra(COMING_FROM_KEY, comingFrom)
            if (comingFrom != COMING_FROM_DIRECT_UNLOCK) {
                starter.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
            }
            context.startActivity(starter)
        }

        private const val COMING_FROM_KEY = "comingFrom"
        const val COMING_FROM_DIRECT_UNLOCK = "comingFromDirectUnlock"
        const val COMING_FROM_QR_SCAN = "comingFromQRScan"
        const val COMING_FROM_NFC = "comingFromNFC"
    }

    private val comingFrom: String?
        get() = intent.extras?.getString(COMING_FROM_KEY)

    private val tileId: String?
        get() = intent.extras?.getString(TILE_ID)

    private val deviceJson: String?
        get() = intent.extras?.getString(DEVICE)

}
